src/dune_rules/module_compilation.ml

```diff
@@ -116,12 +126,17 @@ let build_cm cctx ~precompiled_cmi ~cm_kind (m : Module.t)
   let other_targets, cmt_args =
     match cm_kind with
     | Cmx -> (other_targets, Command.Args.empty)
-    | Cmi | Cmo ->
+    | Cmi | Cmo | Cmj ->
       if Compilation_context.bin_annot cctx then
         let fn =
           Option.value_exn (Obj_dir.Module.cmt_file obj_dir m ~ml_kind)
         in
-        (fn :: other_targets, A "-bin-annot")
+        let bin_annot_arg =
+          match cm_kind with
+          | Cmj -> (* -bin-annot not supported by melange *) Command.Args.empty
```

Comment:
You should hoist this case to the `match cm_kind` that's already present above.

Reply:
Melange support for `bin-annot` was fixed recently, so inner `match` can go away altogether: https://github.com/jchavarri/dune/pull/1/commits/5596594dd7ca166a05db89d43f031138f46b61fb.